README.md

```diff
@@ -36,21 +36,74 @@ There are three metadata files:
-
-The documentation uses the key "#" with a list of strings as the value.
-These strings document how the problem readme (`description.md`) is generally
-interpreted in test programs across different languages.  In addition to a
-mainstream implementation path, this information can also document significant
-variations.
-
-Each test case has the the following keys:
-- description: which will be used to name each generated test
-  - The description should not simply explain **what** each case is (that is redundant information)
-  - The description should explain **why** each case is there. For example, what kinds of implementation mistakes might this case help us find?
-- 'variable names': one or more variable names with values which will be passed to the solution method
-- expected: the expected result
+program.  The file format is described in `canonical-schema.json`, but it
+is easier to understand with a example:
+
+```json
+{ "exercise": "foobar"
+, "version" : "1.0.0"
+, "comments":
+    [ " Comments are always optional and can be used almost anywhere.      "
+    , "                                                                    "
+    , " They usually document how the exercise's readme ('description.md') "
+    , " is generally interpreted in test programs across different         "
+    , " languages.                                                         "
+    , "                                                                    "
+    , " In addition to a mainstream implementation path, this information  "
+    , " can also document significant variations.                          "
+    ]
+, "cases":
+    [ { "comments":
+          [ " A test case must have a 'description' and a 'property'.  "
+          , " Anything else is optional.                               "
+          , "                                                          "
+          , " The 'property' is a string in lowerCamelCase identifying "
+          , " the type of test, but most of the times it is just the   "
+          , " name of a function being tested.                         "
+          , "                                                          "
+          , " Test cases can have any number of additional keys, and   "
+          , " most of them also have an 'expected' one, defining the   "
+          , " value a test should return.                              "
+          ]
+      , "description": "Foo'ing a name returns it reversed"
+      , "property"   : "foo"
+      , "input"      : "lion"
+      , "expected"   : "noil"
+      }
+    , { "description": "Bar'ing a name returns it duplicated"
+      , "property"   : "bar"
+      , "input"      : "Oink"
+      , "expected"   : "OinkOink"
+      }
+    , { "comments":
+          [ " Test cases can be arbitrarily grouped with a description "
+          , " to make organization easier.                             "
+          ]
+      , "description": "Abnormal input: empty strings and numbers"
```

Comment:
Just for the sake of discussion, I found a few possible objection to adding an `input` object:

- We don't really need it. All additional keys in a test case are already implicitly inputs to the test.
- It decreases human-readability, adding an unneeded `input` key and increasing one level of nesting.
- Google JSON Style Guide recommends avoiding unneeded objects.

Following the same reasoning, it would also make more sense to have either a `expected` or an `error`. Maybe we shouldn't have put the `error` inside `expected`.